packages/grid/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts

```diff
@@ -522,7 +515,8 @@ export const useGridCellEditing = (
     }
   }, [cellModesModelProp, updateCellModesModel]);
 
-  React.useEffect(() => {
+  // Run this effect synchronously so that the keyboard event can impact the yet-to-be-rendered input.
```

Comment:
This feels a bit risky, but it works better than other solutions we had before 😅

It also solves https://github.com/mui/mui-x/issues/10582. I tried to cover it with a unit test but failed. I'll try a different approach.

Reply:
I've added e2e tests, but some of them are failing in Firefox.
Investigating...

Reply:
Ended up with a workaround https://github.com/mui/mui-x/pull/9205/commits/28c4ad727d4eff51f8e7b852b5286a0a4bf99f9b
It's ugly, but I didn't want to spend more time investigating it since it passes in other browsers